app/estimate_square_size.go

```diff
@@ -0,0 +1,242 @@
+	coretypes "github.com/tendermint/tendermint/types"
+)
+
+// prune removes txs until the set of txs will fit in the square of size
+// squareSize. It assumes that the currentShareCount is accurate. This function
+// is far from optimal becuse accurately knowing how many shares any given
+// set of transactions and its message takes up in a data square that is following the
+// non-interactive default rules requires recalculating the entire square.
+// TODO: include the padding used by each msg when counting removed shares
+func prune(txConf client.TxConfig, txs []*parsedTx, currentShareCount, squareSize int) parsedTxs {
+	maxShares := squareSize * squareSize
+	if maxShares >= currentShareCount {
+		return txs
+	}
+	goal := currentShareCount - maxShares
+
+	removedContiguousShares := 0
+	contigBytesCursor := 0
+	removedMessageShares := 0
+	removedTxs := 0
+
+	// adjustContigCursor checks if enough contiguous bytes have been removed
+	// inorder to tally total contiguous shares removed
+	adjustContigCursor := func(l int) {
+		contigBytesCursor += l + shares.DelimLen(uint64(l))
+		if contigBytesCursor >= consts.TxShareSize {
+			removedContiguousShares += (contigBytesCursor / consts.TxShareSize)
+			contigBytesCursor = contigBytesCursor % consts.TxShareSize
+		}
+	}
+
+	for i := len(txs) - 1; (removedContiguousShares + removedMessageShares) < goal; i-- {
+		// this normally doesn't happen, but since we don't calculate the number
+		// of padded shares also being removed, its possible to reach this value
+		// should there be many small messages, and we don't want to panic.
+		if i < 0 {
+			break
+		}
+		removedTxs++
+		if txs[i].msg == nil {
+			adjustContigCursor(len(txs[i].rawTx))
+			continue
+		}
+
+		removedMessageShares += shares.MsgSharesUsed(len(txs[i].msg.GetMessage()))
+		// we ignore the error here, as if there is an error malleating the tx,
+		// then it we need to remove it anyway and will not end up contributing
+		// bytes to the square anyway.
+		_ = txs[i].malleate(txConf, uint64(squareSize))
+		adjustContigCursor(len(txs[i].malleatedTx) + appconsts.MalleatedTxBytes)
+	}
+
+	return txs[:len(txs)-(removedTxs)]
+}
+
+// calculateCompactShareCount calculates the exact number of compact shares used.
+func calculateCompactShareCount(txs []*parsedTx, evd core.EvidenceList, squareSize int) int {
+	txSplitter, evdSplitter := shares.NewContiguousShareSplitter(consts.TxNamespaceID), shares.NewContiguousShareSplitter(consts.EvidenceNamespaceID)
+	var err error
+	msgSharesCursor := len(txs)
+	for _, tx := range txs {
+		rawTx := tx.rawTx
+		if tx.malleatedTx != nil {
+			rawTx, err = coretypes.WrapMalleatedTx(tx.originalHash(), uint32(msgSharesCursor), tx.malleatedTx)
+			// we should never get to this point, but just in case we do, we
+			// catch the error here on purpose as we want to ignore txs that are
+			// invalid (cannot be wrapped)
+			if err != nil {
+				continue
+			}
```

Comment:
that's a really good point :thinking: 

the original reason was if someone somehow found how, putting that tx in the mempool would cause all block producers to panic when attempting to create a block with it. In theory this should get caught since we have already decoded the transaction and are just wrapping it in a type ourselves, but since protobuf optimizes for speed over safety (wasn't designed to unmarshal things from byzantine actors) I'm paranoid. Ignoring it would be the safer thing, but I can see how that would be difficult to track down.

perhaps we should log it?

Reply:
Actually, just decided to panic here. I think you're correct that this is best thing to do in this case. Not only just because it will be clear as day, but I think weirder bugs would occur if this did happen, like blocks getting rejected becuase a non-wrapped PFD could make it in the block.